src/main/java/com/google/cloud/spanner/pgadapter/statements/IntermediateStatement.java

```diff
@@ -240,85 +299,191 @@ public Exception getException() {
         break;
       case NO_RESULT:
-        this.hasMoreData = false;
-        this.statementResult = null;
+        this.hasMoreData[index] = false;
         break;
       default:
         throw SpannerExceptionFactory.newSpannerException(
             ErrorCode.INTERNAL, "Unknown or unsupported result type: " + result.getResultType());
     }
   }
 
-  protected void updateBatchResultCount(long[] updateCounts) {
-    this.updateCount = 0L;
-    for (int i = 0; i < updateCounts.length; ++i) {
-      this.updateCount += updateCounts[i];
+  protected void updateBatchResultCount(int fromIndex, long[] updateCounts) {
+    for (int index = fromIndex; index < fromIndex + updateCounts.length; index++) {
+      setUpdateCount(index, updateCounts[index - fromIndex]);
     }
-    this.hasMoreData = false;
-    this.statementResult = null;
   }
 
   /**
    * Clean up and save metadata when an exception occurs.
    *
    * @param e The exception to store.
    */
-  protected void handleExecutionException(SpannerException e) {
-    this.exception = e;
-    this.hasMoreData = false;
-    this.statementResult = null;
+  protected void handleExecutionException(int index, SpannerException e) {
+    this.exceptions[index] = e;
+    this.hasMoreData[index] = false;
+  }
+
+  private void handleExecutionExceptionAndTransactionStatus(int index, SpannerException e) {
+    if (executionStatus == ExecutionStatus.EXPLICIT_TRANSACTION) {
+      connectionHandler.setStatus(ConnectionStatus.TRANSACTION_ABORTED);
+    }
+    handleExecutionException(index, e);
   }
 
   /** Execute the SQL statement, storing metadata. */
   public void execute() {
-    this.executed = true;
-    try {
-      if (statements.size() > 1) {
-        // TODO: Clean this up a little once the statement parsing in the client library has been
-        // released: https://github.com/googleapis/java-spanner/pull/1690
-        // Also, the restriction that mixed batches are not allowed will be removed in a future PR.
-        if (resultType == ResultType.UPDATE_COUNT) {
-          connection.startBatchDml();
-        } else if (resultType == ResultType.NO_RESULT) {
-          connection.startBatchDdl();
-        } else {
-          throw SpannerExceptionFactory.newSpannerException(
-              ErrorCode.INVALID_ARGUMENT, "Statement type is not supported for batching");
-        }
-        for (String stmt : statements) {
-          connection.execute(Statement.of(stmt));
-        }
-        long[] updateCounts = connection.runBatch();
-        updateBatchResultCount(updateCounts);
+    executed = true;
+    if (connection.isInTransaction()) {
+      executionStatus = ExecutionStatus.EXPLICIT_TRANSACTION;
+    } else {
+      executionStatus = ExecutionStatus.AUTOCOMMIT;
+    }
+    int index = 0;
+    while (index < getStatementCount()) {
+      if (connectionHandler.getStatus() == ConnectionStatus.TRANSACTION_ABORTED
+          && !"ROLLBACK".equals(getCommand(index))
```

Comment:
The checks for rollback/commit must be case-insensitive (the best way is probably to fill the command list with all upper-case commands, as PostgreSQL will always return the command tags in upper case).

Reply:
The commands are forced to be upper-case when they are parsed.